java/sage/media/bluray/BluRayFile.java

```diff
@@ -0,0 +1,496 @@
+
+    if (bytesRead < count)
+    {
+      ensureProperFile(false);
+      bytesRead += sageFileChannel.transferTo(count - bytesRead, target);
+    }
+
+    return bytesRead;
+  }
+
+  @Override
+  public void openFile(String name) throws IOException
+  {
+    sageFileChannel.openFile(name);
+  }
+
+  @Override
+  public void openFile(File file) throws IOException
+  {
+    sageFileChannel.openFile(file);
+  }
+
+  /**
+   * Unsupported method. This channel can only be read only.
+   *
+   * @throws IOException throws channel is read only.
+   */
+  @Override
+  public long transferFrom(ReadableByteChannel src, long position, long count) throws IOException
+  {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Unsupported method. This channel can only be read only.
+   *
+   * @throws IOException throws channel is read only.
+   */
+  @Override
+  public long transferFrom(ReadableByteChannel src, long count) throws IOException
+  {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Unsupported method. This channel can only be read only.
+   *
+   * @throws IOException throws channel is read only.
+   */
+  @Override
+  public int write(ByteBuffer src, long position) throws IOException
+  {
+    throw new NonWritableChannelException();
+  }
+
+  /**
+   * Unsupported method.
+   *
+   * @throws IOException throws method is unsupported.
+   */
+  @Override
+  public MappedByteBuffer map(MapMode mode, long position, long size) throws IOException
+  {
+    throw new IOException("Method is unsupported.");
+  }
+
+  /**
+   * Unsupported method.
+   *
+   * @throws IOException throws method is unsupported.
+   */
+  @Override
+  public FileLock lock(long position, long size, boolean shared) throws IOException
+  {
+    throw new IOException("Method is unsupported.");
+  }
+
+  /**
+   * Unsupported method.
+   *
+   * @throws IOException throws method is unsupported.
+   */
+  @Override
+  public FileLock tryLock(long position, long size, boolean shared) throws IOException
+  {
+    throw new IOException("Method is unsupported.");
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException
+  {
+    int readBytes = sageFileChannel.read(dst, position - fileOffsets[currFileIndex]);
+
+    if (dst.hasRemaining())
+    {
+      ensureProperFile(false);
+      if (readBytes == -1)
+        readBytes = 0;
+      readBytes += sageFileChannel.read(dst, position + readBytes - fileOffsets[currFileIndex]);
+    }
+
+    return readBytes;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException
+  {
+    int readBytes = sageFileChannel.read(dst);
+
+    if (dst.hasRemaining())
+    {
+      ensureProperFile(false);
+      if (readBytes == -1)
+        readBytes = 0;
+      readBytes += sageFileChannel.read(dst);
+    }
+
+    return readBytes;
+  }
+
+  @Override
+  public int readUnsignedByte() throws IOException
+  {
+    return sageFileChannel.readUnsignedByte();
```

Comment:
This should have an ensureProperFile(false) at the beginning...and actually all read methods, even non-absolute ones need that in case they are called when the current file is at EOF.